### PR TITLE
spec v0.6

### PR DESCRIPTION
**Pull request by wemeetagain**
- update state state transition logic to 0.6
- remove stale stateTransitionHelpers, add new ones
- move stateTransitionHelpers megafile into stateTransition/util/* files
- reorganize stateTransitionHelpers tests
- move src/helpers to src/util
  - add utils from stateTransitionHelpers
- lint --fix

### CODE REVIEW

**Review by mpetrunic — DISMISSED** [review]
wow, lgtm

**Comment by GregTheGreek** [comment]
FYI -- Still reviewing! It's just you know, big

File: src/chain/index.ts [hunk 1]
```diff
@@ -75,11 +76,11 @@ export class BeaconChain extends EventEmitter {
```
Comment (GregTheGreek) [inline src/chain/index.ts:76]: not part of this PR but `isValidBlock` could be reduced to one line

Comment (wemeetagain) [reply src/chain/index.ts:76]: isValidBlock is still a WIP, but it's intended to be this part of the fork choice rule spec:

 > For a beacon block, block, to be processed by a node, the following conditions must be met:
> *   The parent block with root block.previous_block_root has been processed and accepted.
> *   An Ethereum 1.0 block pointed to by the state.latest_eth1_data.block_hash has been processed and accepted.
> *   The node's Unix time is greater than or equal to state.genesis_time + block.slot * SECONDS_PER_SLOT.

https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_fork-choice.md#beacon-chain-processing

File: src/chain/stateTransition/block/attestations.ts [hunk 2]
```diff
@@ -1,100 +1,85 @@
...
+  const currentEpoch = getCurrentEpoch(state);
+  const previousEpoch = getPreviousEpoch(state);
+  const data = attestation.data;
+  const minSlot =  currentEpoch > GENESIS_EPOCH
+    ? state.slot - SLOTS_PER_EPOCH
+    : GENESIS_SLOT;
+  assert(minSlot <= data.slot && data.slot <= state.slot - MIN_ATTESTATION_INCLUSION_DELAY);
+
+  // Check target epoch, source epoch, and source root
+  const targetEpoch = slotToEpoch(attestation.data.slot);
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/attestations.ts]: should use `data.slot` since it's already a const on line 34

File: src/chain/stateTransition/block/attestations.ts [hunk 3]
```diff
@@ -1,100 +1,85 @@
...
+  assert(
+    currentEpoch === targetEpoch ||
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/attestations.ts]: ~Not sure if we want this, it will block us from being able to get to line 73~

The python version is asserting everything against itself.

Comment (GregTheGreek) [reply src/chain/stateTransition/block/attestations.ts]: Something like

`assert(["target_epoch","current_epoch"].includes("test","test1"))`

File: src/chain/stateTransition/block/deposits.ts [hunk 4]
```diff
@@ -1,38 +1,90 @@
...
+    // Increase balance by deposit amount
+    const index = validatorPubkeys.indexOf(pubkey);
+    increaseBalance(state, index, amount);
+  }
+}
 
 export default function processDeposits(state: BeaconState, block: BeaconBlock): void {
-  assert(block.body.deposits.length <= MAX_DEPOSITS);
-  // TODO: add logic to ensure that deposits from 1.0 chain are processed in order
+  assert(block.body.deposits.length === Math.min(MAX_DEPOSITS, block.body.eth1Data.depositCount - state.depositIndex));
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/deposits.ts]: I believe we want to be checking the latest_eth1Data from state:

`assert(block.body.deposits.length === Math.min(MAX_DEPOSITS, state.latestEth1Data.depositCount - state.depositIndex));`

File: src/chain/stateTransition/block/eth1Data.ts [hunk 5]
```diff
@@ -1,22 +1,25 @@
...
+export default function processEth1Data(state: BeaconState, block: BeaconBlock): void {
+  const blockEth1Data = block.body.eth1Data;
+  state.eth1DataVotes.push(blockEth1Data);
+  const serializedBlockEth1Data = serialize(blockEth1Data, Eth1Data);
+  let occurances = 0;
+  state.eth1DataVotes.forEach((eth1Data) => {
+    if (serialize(eth1Data, Eth1Data).equals(serializedBlockEth1Data)) {
+      occurances++;
+    }
+  });
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/eth1Data.ts:21]: I'm curious if serializing all the votes will cause performance issues?

Might be worth making a note this could be a bottlneck area when we reach a few 100k validators

File: src/chain/stateTransition/block/proposerSlashings.ts [hunk 6]
```diff
@@ -1,55 +1,60 @@
...
-} from "../../helpers/validatorStatus";
-
-import {blsVerify} from "../../../stubs/bls";
+} from "../util";
+
+
+export function processProposerSlashing(state, proposerSlashing: ProposerSlashing): void {
+  const proposer = state.validatorRegistry[proposerSlashing.proposerIndex];
+  // Verify that the epoch is the same
+  assert(proposerSlashing.header1.slot === proposerSlashing.header2.slot);
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/proposerSlashings.ts]: You need to wrap these in `slotToEpoch` otherwise you're comparing a slot which mihgt not be in the same epoch

File: src/chain/stateTransition/block/transfers.ts [hunk 7]
```diff
@@ -11,52 +10,68 @@ import {
...
+    transfer.signature,
+    getDomain(state, Domain.TRANSFER, slotToEpoch(transfer.slot)),
+  ));
+  // Process the transfer
+  decreaseBalance(state, transfer.sender, transfer.amount.add(transfer.fee));
+  increaseBalance(state, transfer.recipient, transfer.amount);
+  increaseBalance(state, getBeaconProposerIndex(state), transfer.fee);
+  // Verify balances are not dust
+  assert(!(state.balances[transfer.sender].ltn(MIN_DEPOSIT_AMOUNT)));
+  assert(!(state.balances[transfer.recipient].ltn(MIN_DEPOSIT_AMOUNT)));
```
Comment (GregTheGreek) [inline src/chain/stateTransition/block/transfers.ts]: I don't think we have a check for `0 < state.balances[transfer...]` anywhere else, so it may be worth doing that

File: src/chain/stateTransition/epoch/crosslinks.ts [hunk 8]
```diff
@@ -1,32 +1,33 @@
...
-  state: BeaconState,
-  previousEpoch: Epoch,
-  nextEpoch: Epoch,
-  previousEpochAttestations: PendingAttestation[],
-  currentEpochAttestations: PendingAttestation[]): void {
+import {getWinningCrosslinkAndAttestingIndices} from "./util";
+
+
+export function processCrosslinks(state: BeaconState): void {
+  const currentEpoch = getCurrentEpoch(state);
```
Comment (GregTheGreek) [inline src/chain/stateTransition/epoch/crosslinks.ts:17]: missing `state.previousCrosslinks = state.currentCrosslinks.map((c: crossLink) => c));

File: src/chain/stateTransition/epoch/justification.ts [hunk 9]
```diff
@@ -1,41 +1,62 @@
...
-    state.finalizedEpoch = state.justifiedEpoch;
+  // The 1st/2nd/3rd most recent epochs are all justified, the 1st using the 3rd as source
+  if (state.justificationBitfield.shrn(0).modn(8) === 0b111 && state.currentJustifiedEpoch === currentEpoch - 2) {
+    state.finalizedEpoch = oldCurrentJustifiedEpoch;
+    state.finalizedRoot = finalizedRoot;
   }
-  if (state.justificationBitfield.shrn(0).modn(4) === 0b11 && state.justifiedEpoch === previousEpoch) {
-    state.finalizedEpoch = state.justifiedEpoch;
+  // The 1st/2nd most recent epochs are both justified, the 1st using the 2nd as source
+  if (state.justificationBitfield.shrn(0).modn(4) === 0b11 && state.currentJustifiedEpoch === currentEpoch - 1) {
```
Comment (GregTheGreek) [inline src/chain/stateTransition/epoch/justification.ts]: For all `state.<current || previous>JustifiedEpoch` theres a constant on line 13 and 14 you should probably use

File: src/chain/stateTransition/epoch/util.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      if (b.balance.gt(a.balance)) {
+        return b;
+      } else if (b.balance.eq(a.balance)) {
+        if (deserialize(b.crosslink.crosslinkDataRoot, "uint256").gt(deserialize(a.crosslink.crosslinkDataRoot, "uint256"))) {
+          return b;
+        }
+      }
+      return a;
+    }).crosslink;
+  return [winningCrosslink, getUnslashedAttestingIndices(state, getAttestationsFor(winningCrosslink))];
```
Comment (GregTheGreek) [inline src/chain/stateTransition/epoch/util.ts:111]: This one took me a minute A+

File: src/chain/stateTransition/util/attestation.ts [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+/**
+ * Returns the ith bit in bitfield
+ * @param {bytes} bitfield
+ * @param {number} i
+ * @returns {number}
+ */
+export function getBitfieldBit(bitfield: bytes, i: number): number {
+  const bit = i % 8;
+  const byte = intDiv(i,  8);
+  return (bitfield[byte] >> bit) & 1;
```
Comment (GregTheGreek) [inline src/chain/stateTransition/util/attestation.ts:60]: curious what the `& 1` does here?

File: src/chain/stateTransition/util/attestation.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+  // TO BE REMOVED IN PHASE 1
+  if (custodyBit0Indices.length > 0) {
+    return false;
+  }
+
+  const totalAttestingIndices = custodyBit0Indices.length + custodyBit1Indices.length;
+  if (!(1 <= totalAttestingIndices && totalAttestingIndices <= MAX_INDICES_PER_ATTESTATION)) {
+    return false;
+  }
+
```
Comment (GregTheGreek) [inline src/chain/stateTransition/util/attestation.ts:128]: whitespace

File: src/chain/stateTransition/util/epoch.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Return the previous epoch of the given state.
+ * @param {BeaconState} state
+ * @returns {Epoch}
+ */
+export function getPreviousEpoch(state: BeaconState): Epoch {
+  const currentEpoch = getCurrentEpoch(state);
+  if (currentEpoch == GENESIS_EPOCH) {
+    return GENESIS_EPOCH;
```
Comment (GregTheGreek) [inline src/chain/stateTransition/util/epoch.ts:30]: I would `===` this statement, otherwise conform to the spec and return `currentEpoch`

File: src/constants/constants.ts [hunk 14]
```diff
@@ -1,65 +1,72 @@
...
 export const LATEST_ACTIVE_INDEX_ROOTS_LENGTH = 2 ** 13; // epochs || ~36 days
 export const LATEST_SLASHED_EXIT_LENGTH = 2 ** 13; // epochs || ~36 days
 
 // Reward and penalty quotients
 export const BASE_REWARD_QUOTIENT = 2 ** 5; // 32
-export const WHISTLEBLOWER_REWARD_QUOTIENT = 2 ** 9; // 512
-export const ATTESTATION_INCLUSION_REWARD_QUOTIENT = 2 ** 3; // 8
+export const WHISTLEBLOWING_REWARD_QUOTIENT = 2 ** 9; // 512
+export const PROPOSER_REWARD_QUOTIENT = 2 ** 3; // 8
 export const INACTIVITY_PENALTY_QUOTIENT = 2 ** 24; // 16,777,216
```
Comment (GregTheGreek) [inline src/constants/constants.ts]: Should be `2**25`

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
I combed through pretty much everything, I found a few small things, i think one or two critical issues. Otherwise all bueno!

Going through this I think we can open a neat bounty, to have someone add types to all our declared `consts` etc... sometimes it was tough to figure out what we are iterating over

I'm also pretty sure I found 18 issues but I accidentally force closed my browser and it says 14 now ...

File: src/chain/stateTransition/util/attestation.ts (hunk 11 above)
Comment (wemeetagain) [reply src/chain/stateTransition/util/attestation.ts:60]: Its a bitwise AND operation.
Checks the least significant bit, returning 0 if 0, 1 if 1. Its basically equivalent to doing n % 2.

File: src/chain/stateTransition/block/eth1Data.ts (hunk 5 above)
Comment (wemeetagain) [reply src/chain/stateTransition/block/eth1Data.ts:21]: good question. We'll definitely see when we start benchmarking.

**Review by GregTheGreek — APPROVED** [review]

**Comment by GregTheGreek** [comment]
LGTM Let's just break things from here

**Review by mpetrunic — APPROVED** [review]
